Realm/RLMValue.mm

```diff
@@ -0,0 +1,134 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2020 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#import "RLMValue.h"
+#import "RLMUtil.hpp"
+
+#pragma mark NSData
+
+@implementation NSData (RLMValue)
+
+- (RLMPropertyType)rlm_valueType {
+    return RLMPropertyTypeData;
+}
+
+@end
+
+#pragma mark NSDate
+
+@implementation NSDate (RLMValue)
+
+- (RLMPropertyType)rlm_valueType {
+    return RLMPropertyTypeDate;
+}
+
+@end
+
+#pragma mark NSNumber
+
+@implementation NSNumber (RLMValue)
+
+- (RLMPropertyType)rlm_valueType {
+    if (numberIsBool(self)) {
+        return RLMPropertyTypeBool;
+    }
+    else if (numberIsInteger(self)) {
+        return RLMPropertyTypeInt;
+    }
+    else if (*@encode(float) == [self objCType][0]) {
+        return RLMPropertyTypeFloat;
+    }
+    else if (*@encode(double) == [self objCType][0]) {
+        return RLMPropertyTypeDouble;
+    }
+    else {
+        @throw RLMException(@"Unknown numeric type on type RLMValue.");
+    }
+}
+
+@end
+
+#pragma mark NSNull
+
+@implementation NSNull (RLMValue)
+
+- (RLMPropertyType)rlm_valueType {
+    return RLMPropertyTypeAny;
```

Comment:
This doesn't seem correct

Reply:
RLMPropertyType is an enum, I think returning nil would not be the correct approach.

Reply:
I agree that would shouldn't return nil. I guess this should either return what you have here or `-1` to imply that it's an unknown type.

Reply:
I think `Any` is probably the least-bad option here. The options basically are:
1. Make the return type optional.
2. Add Null to RLMPropertyType
3. Add RLMMixedType
4. Just return Any

1 would be ideal but isn't really possible in obj-c. 2 is clearly incorrect as Null isn't a valid property type. 3 is probably the most correct but makes everything a lot more awkward. 4 is incorrect in a less problematic way and probably isn't an actual problem in practice.